basic02-prog-by-name/xdp_loader.c

```diff
@@ -59,106 +59,76 @@ static const struct option_wrapper long_options[] = {
-	 */
-	struct bpf_prog_load_attr prog_load_attr = {
-		.prog_type	= BPF_PROG_TYPE_XDP,
-		.ifindex	= ifindex,
-	};
-	prog_load_attr.file = filename;
-
-	/* Use libbpf for extracting BPF byte-code from BPF-ELF object, and
-	 * loading this into the kernel via bpf-syscall
-	 */
-	err = bpf_prog_load_xattr(&prog_load_attr, &obj, &first_prog_fd);
-	if (err) {
-		fprintf(stderr, "ERR: loading BPF-OBJ file(%s) (%d): %s\n",
-			filename, err, strerror(-err));
-		return NULL;
-	}
 
-	/* Notice how a pointer to a libbpf bpf_object is returned */
-	return obj;
-}
 
 /* Lesson#2: This is a central piece of this lesson:
  * - Notice how BPF-ELF obj can have several programs
  * - Find by sec name via: bpf_object__find_program_by_title()
  */
-struct bpf_object *__load_bpf_and_xdp_attach(struct config *cfg)
+struct xdp_program *__load_bpf_and_xdp_attach(struct config *cfg)
 {
 	/* In next assignment this will be moved into ../common/ */
-	struct bpf_program *bpf_prog;
-	struct bpf_object *bpf_obj;
-	int offload_ifindex = 0;
 	int prog_fd = -1;
 	int err;
 
+	DECLARE_LIBBPF_OPTS(bpf_object_open_opts, opts);
+	DECLARE_LIBXDP_OPTS(xdp_program_opts, xdp_opts, 0);
+
+	xdp_opts.open_filename = cfg->filename;
+	xdp_opts.prog_name = cfg->progsec;
+	xdp_opts.opts = &opts;
+
 	/* If flags indicate hardware offload, supply ifindex */
-	if (cfg->xdp_flags & XDP_FLAGS_HW_MODE)
-		offload_ifindex = cfg->ifindex;
+	/* if (cfg->xdp_flags & XDP_FLAGS_HW_MODE) */
+	/* 	offload_ifindex = cfg->ifindex; */
 
-	/* Load the BPF-ELF object file and get back libbpf bpf_object */
-	bpf_obj = __load_bpf_object_file(cfg->filename, offload_ifindex);
-	if (!bpf_obj) {
-		fprintf(stderr, "ERR: loading file: %s\n", cfg->filename);
+	struct xdp_program *prog = xdp_program__create(&xdp_opts);
+	err = libxdp_get_error(prog);
+	if (err) {
+		char errmsg[1024];
+		libxdp_strerror(err, errmsg, sizeof(errmsg));
+		fprintf(stderr, "ERR: loading program: %s\n", errmsg);
 		exit(EXIT_FAIL_BPF);
 	}
+
 	/* At this point: All XDP/BPF programs from the cfg->filename have been
 	 * loaded into the kernel, and evaluated by the verifier. Only one of
 	 * these gets attached to XDP hook, the others will get freed once this
 	 * process exit.
 	 */
 
-	/* Find a matching BPF prog section name */
-	bpf_prog = bpf_object__find_program_by_title(bpf_obj, cfg->progsec);
-	if (!bpf_prog) {
-		fprintf(stderr, "ERR: finding progsec: %s\n", cfg->progsec);
-		exit(EXIT_FAIL_BPF);
-	}
-
-	prog_fd = bpf_program__fd(bpf_prog);
-	if (prog_fd <= 0) {
-		fprintf(stderr, "ERR: bpf_program__fd failed\n");
-		exit(EXIT_FAIL_BPF);
-	}
-
 	/* At this point: BPF-progs are (only) loaded by the kernel, and prog_fd
 	 * is our select file-descriptor handle. Next step is attaching this FD
 	 * to a kernel hook point, in this case XDP net_device link-level hook.
 	 */
-	err = xdp_link_attach(cfg->ifindex, cfg->xdp_flags, prog_fd);
+	err = xdp_program__attach(prog, cfg->ifindex, XDP_MODE_SKB, 0);
```

Comment:
just the mode - we may want to keep this as a flag

Reply:
As previous, fixing in v2 branch.